### PR TITLE
rgw: Fixes and Improvements for Datalog

### PR DESCRIPTION
**Pull request by adamemerson**
- Allow online changes in log format in situ
- Don't hammer the OSD on startup
- A few fixes and silencing of spurious errors

### CODE REVIEW

**Comment by adamemerson** [comment]
@cbodley Please give this a lookover for sanity if you can

**Review by mattbenjamin — APPROVED** [review]
looks good to me, but I don't know lots of fifo details

File: src/rgw/rgw_admin.cc [hunk 1]
```diff
@@ -244,6 +244,8 @@ void usage()
   cout << "  datalog list               list data log\n";
   cout << "  datalog trim               trim data log\n";
   cout << "  datalog status             read data log status\n";
+  cout << "  datalog type               change datalog type\n";
```
Comment (cbodley) [inline src/rgw/rgw_admin.cc:247]: maybe `change datalog type to --log-type={fifo,omap}`

File: src/rgw/rgw_admin.cc [hunk 2]
```diff
@@ -244,6 +244,8 @@ void usage()
...
+  cout << "  datalog prune              prune empty datalog generations\n";
```
Comment (cbodley) [inline src/rgw/rgw_admin.cc:248]: when/why would a user need to run this? can trimming not handle this automatically?

File: src/rgw/rgw_datalog.h [hunk 3]
```diff
@@ -303,6 +309,8 @@ class RGWDataChangesBE : public boost::intrusive_ref_counter<RGWDataChangesBE> {
   virtual int trim(int index, std::string_view marker,
 		   librados::AioCompletion* c) = 0;
   virtual std::string_view max_marker() const = 0;
+  // 1 on empty, 0 on non-empty, negative o error.
+  virtual int is_empty() = 0;
```
Comment (cbodley) [inline src/rgw/rgw_datalog.h:313]: consider making this const?

File: src/common/dynarray.h [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+#include "include/uses_allocator.h"
+
+/// The problem:
+///     - Non-movable types
+///     - Non-default-constructible types
+///     - Entries whose constructor parameters depend on the index
```
[the remark this answers is not in the document]
Comment (adamemerson) [reply src/common/dynarray.h:22]: I don't really care for the way tiny-vector does its constructors, having looked at it, though I can use it, I guess.

[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/common/dynarray.h:22]: maybe converging on a different container implementation--if indeed possible, and if indeed better for purpose--could be deferred to refactoring

File: src/rgw/rgw_admin.cc (hunk 2 above)
Comment (adamemerson) [reply src/rgw/rgw_admin.cc:248]: I put it in for testing, basically. I wanted to make sure we were getting rid of the right generations.

File: src/rgw/rgw_datalog.h (hunk 3 above)
Comment (adamemerson) [reply src/rgw/rgw_datalog.h:313]: It makes a bunch of RADOS calls and stuff, though I could make the IoCtx mutable.

File: src/rgw/rgw_datalog.cc [hunk 5]
```diff
@@ -653,20 +684,105 @@ int RGWDataChangesLog::list_entries(int max_entries,
...
+  std::unique_lock l(m);
+  const auto head_gen = (end() - 1)->second->gen_id;
+  const auto tail_gen = begin()->first;
+  if (target_gen < tail_gen) return 0;
+  auto r = 0;
+  for (auto i = lower_bound(0);
+       i != end() && i->first <= target_gen && i->first <= head_gen && r >= 0;
+       i = upper_bound(i->first)) {
+    auto be = i->second;
+    l.unlock();
```
[the remark this answers is not in the document]
Comment (adamemerson) [reply src/rgw/rgw_datalog.cc:790]: Thank you, forgot about the thing we're pointing to being at risk of deletion.

File: src/common/dynarray.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (cbodley) [reply src/common/dynarray.h:22]: let's use the existing container for now. if you have ideas to improve on that, let's revisit that later in a way that will benefit existing users of `tiny_vector`

File: src/rgw/rgw_admin.cc (hunk 2 above)
Comment (cbodley) [reply src/rgw/rgw_admin.cc:248]: i'm fine with admin commands that aid testing, but not all of those should be documented for users. if it's unnecessary in normal operation, and more likely to cause problems than solve them, then i'd prefer not to document it

File: src/rgw/rgw_datalog.h (hunk 3 above)
Comment (cbodley) [reply src/rgw/rgw_datalog.h:313]: ok nevermind!

File: src/rgw/rgw_datalog.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (adamemerson) [reply src/rgw/rgw_datalog.cc:790]: Fixed. Thank you.

File: src/rgw/rgw_admin.cc (hunk 1 above)
Comment (adamemerson) [reply src/rgw/rgw_admin.cc:247]: Done.

File: src/rgw/rgw_admin.cc (hunk 2 above)
Comment (adamemerson) [reply src/rgw/rgw_admin.cc:248]: Done.

File: src/common/dynarray.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (adamemerson) [reply src/common/dynarray.h:22]: Done

**Comment by adamemerson** [comment]
Jenkins, dear machine, retest this, please.

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
jenkins test make check

**Comment by adamemerson** [comment]
@cbodley https://pulpito.ceph.com/aemerson-2021-03-29_21:13:59-rgw-wip-datalog-genbacking-distro-basic-smithi/6010155/

Does this look reasonable? I think the `test_suspended_delete_marker_full_sync` failure is a known issue?

**Comment by cbodley** [comment]
> @cbodley https://pulpito.ceph.com/aemerson-2021-03-29_21:13:59-rgw-wip-datalog-genbacking-distro-basic-smithi/6010155/

looks good!

> Does this look reasonable? I think the `test_suspended_delete_marker_full_sync` failure is a known issue?

https://tracker.ceph.com/issues/45620

**Comment by cbodley** [comment]
do we have a tracker issue for this work? we'll need it on pacific too right

**Comment by adamemerson** [comment]
Does now. I'll merge it in and make a backport.

https://tracker.ceph.com/issues/50104

**Pull request merged**